src/diff/changes/apiCollectionChanges.ts

```diff
@@ -52,4 +54,112 @@ export class ApiCollectionChanges {
   hasErrors(): boolean {
     return Object.keys(this.errored).length > 0;
   }
+
+  /**
+   * Gets the number of changes in each category
+   */
+  getCategorizedChangeSummary(): Record<RuleCategory, number> {
+    const entries = Object.values(RuleCategory).map((c) => [c, 0]);
+    const summaries = Object.values(this.changed).map((api) => {
+      return api.getCategorizedChangeSummary();
+    });
+    return _.mergeWith(
+      _.fromPairs(entries),
+      ...summaries,
+      (a: number | undefined, b: number): number => (a || 0) + b
+    );
+  }
+
+  /**
+   * Format the changes as a string for printing to console
+   *
+   * @param extraIndent - Number of levels the headings should be nested
+   */
+  toString(extraIndent = 0): string {
+    // Ensure passed value is a non-negative number
+    extraIndent = Math.max(0, extraIndent) || 0;
+    /**
+     * Indents the first non-empty line by the sum of the amounts passed to this
+     * function and the outer function. Also appends a newline.
+     * @param baseIndent Base amount to indent
+     * @param str String to indent
+     * @returns The indented string
+     */
+    const indent = (baseIndent: number, str: string): string =>
+      str.replace(/^\n*/, "$&" + " ".repeat(baseIndent + extraIndent)) + "\n";
+
+    let out = "";
+
+    const changedEntries = Object.entries(this.changed).filter(([, api]) => {
+      return api.hasCategorizedChanges();
+    });
+    if (changedEntries.length > 0) {
```

Comment:
A lot of this logic could be simplified by using a template...something like handlebars perhaps.  This has a lot of complex data specific logic here.

Reply:
I considered using handlebars when I first started working on this. I can't even vaguely remember why I decided not to do it, it's a much better idea. :|